File: glTest/src/glTest/framework/ProblemFactory.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package glTest.framework;

import glTest.problems.NullProblem;
import glTest.problems.Problem;
import glTest.solutions.Solution;
import java.util.ArrayList;
import java.util.HashMap;
import glTest.problems.DynamicStreamingProblem;
import glTest.problems.UntexturedObjectsProblem;
import glTest.solutions.null_.NullSolution;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLBufferSubData;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLBufferSubData3Buffers;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLMapPersistent;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLMapPersistent3Buffers;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLMapUnsynchronized;
import glTest.solutions.dynamicStreaming.DynamicStreamingGLMapUnsynchronized3Buffers;
import glTest.solutions.untexturedObjects.bindless.UntexturedObjectsGLBindless;
import glTest.solutions.untexturedObjects.bindlessIndirect.UntexturedObjectsGLBindlessIndirect;
import glTest.solutions.untexturedObjects.bufferRange.UntexturedObjectsGLBufferRange;
import glTest.solutions.untexturedObjects.bufferStorage.UntexturedObjectsGLBufferStorage;
import glTest.solutions.untexturedObjects.drawLoop.UntexturedObjectsGLDrawLoop;
import glTest.solutions.untexturedObjects.dynamicBuffer.UntexturedObjectsGLDynamicBuffer;
import glTest.solutions.untexturedObjects.mapPersistent.UntexturedObjectsGLMapPersistent;
import glTest.solutions.untexturedObjects.mapUnsynchronized.UntexturedObjectsGLMapUnsynchronized;
import glTest.solutions.untexturedObjects.multiDrawBuffer.UntexturedObjectsGLMultiDrawBuffer;
import glTest.solutions.untexturedObjects.texCoord.UntexturedObjectsGLTexCoord;
import glTest.solutions.untexturedObjects.uniform.UntexturedObjectsGLUniform;

/**
 *
 * @author GBarbieri
 */
public class ProblemFactory {

    private ArrayList<Problem> problems = new ArrayList<>();
    private HashMap<String, Solution[]> solutions = new HashMap<>();

    public ProblemFactory() {

        Problem newProbl;

        /**
         * Null.
         */
        newProbl = new NullProblem();

        problems.add(newProbl);
        solutions.put(newProbl.getName(),
                new Solution[]{
                    new NullSolution()});

        /**
         * DynamicStreaming.
         */
        newProbl = new DynamicStreamingProblem();

        problems.add(newProbl);
        solutions.put(newProbl.getName(),
                new Solution[]{
                    new DynamicStreamingGLBufferSubData(),
                    //                    Not working the wanted way
                    //                    new DynamicStreamingGLBufferSubData3Buffers(),
                    //                    new DynamicStreamingGLMapUnsynchronized3Buffers(),
                    //                    new DynamicStreamingGLMapPersistent3Buffers(),
                    new DynamicStreamingGLMapUnsynchronized(),
                    new DynamicStreamingGLMapPersistent()
                });
        /**
         * UntexturedObjects.
         */
        newProbl = new UntexturedObjectsProblem();

        problems.add(newProbl);
        solutions.put(newProbl.getName(),
                new Solution[]{
                    new UntexturedObjectsGLUniform(),
                    new UntexturedObjectsGLDrawLoop(),
                    //                    Bug
                    //                    new UntexturedObjectsGLMultiDraw(true),
                    new UntexturedObjectsGLMultiDrawBuffer(true),
                    new UntexturedObjectsGLMultiDrawBuffer(false),
                    new UntexturedObjectsGLBindless(),
                    new UntexturedObjectsGLBindlessIndirect(), // warning, buffers seen as non-resident
                    new UntexturedObjectsGLBufferRange(),
                    new UntexturedObjectsGLBufferStorage(true),
                    new UntexturedObjectsGLBufferStorage(false),
                    new UntexturedObjectsGLDynamicBuffer(),
                    new UntexturedObjectsGLMapUnsynchronized(),
                    new UntexturedObjectsGLMapPersistent(),
                    new UntexturedObjectsGLTexCoord()
                });
    }

    public ArrayList<Problem> getProblems() {
        return problems;
    }

    public Solution[] getSolutions(Problem problem) {

        assert (problem != null);

        return solutions.get(problem.getName());
    }
}

File: src/main/kotlin/main/glTest/framework/ProblemFactory.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package main.glTest.framework

import main.glTest.problems.*
import main.glTest.solutions.*

/**
 *
 * @author GBarbieri
 */
class ProblemFactory(skipInit: Boolean) {

    val problems = ArrayList<Problem>()
    private val solutions = mutableMapOf<String, Array<Solution>>()

    init {
        // Null
        var newProb: Problem = NullProblem()
        if (skipInit || newProb.init()) {
            if (!skipInit)
                newProb.shutdown()
            problems += newProb
            solutions[newProb.name] = arrayOf<Solution>(NullSolution())
        } else {
            newProb.shutdown()
            throw Error("Unable to create the Null Problem--exiting.")
        }

        // DynamicStreaming
        newProb = DynamicStreamingProblem()
        if (skipInit || newProb.init()) {
            if (!skipInit)
                newProb.shutdown()
            problems += newProb
            solutions[newProb.name] = arrayOf<Solution>(
                    DynamicStreamingGLBufferSubData(),
                    DynamicStreamingGLMapUnsynchronized(),
                    DynamicStreamingGLMapPersistent())
        } else
            newProb.shutdown()

        // Objects
        newProb = ObjectsProblem()
        if (skipInit || newProb.init()) {
            if (!skipInit)
                newProb.shutdown()

            problems += newProb
            solutions[newProb.name] = arrayOf<Solution>(
                    ObjectsGLUniform(),
                    ObjectsGLDrawLoop(),
                    ObjectsGLMultiDraw(true),
                    ObjectsGLMultiDraw(false),
                    ObjectsGLMultiDrawBuffer(true),
                    ObjectsGLMultiDrawBuffer(false),
                    ObjectsGLBindless(),
//                    ObjectsGLBindlessIndirect(),
                    ObjectsGLBufferRange(),
                    ObjectsGLBufferStorage(true),
                    ObjectsGLBufferStorage(false),
                    ObjectsGLDynamicBuffer(),
                    ObjectsGLMapUnsynchronized(),
                    ObjectsGLMapPersistent(),
                    ObjectsGLTexCoord())
        } else
            newProb.shutdown()

        // Textured Quads
        newProb = TexturedQuadsProblem ()
        if (skipInit || newProb.init()) {
            if (!skipInit)
                newProb.shutdown()

            problems += newProb
            solutions[newProb.name] = arrayOf<Solution>(
                    TexturedQuadsGLBindless())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLBindlessMultiDraw())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLNaive())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLNaiveUniform())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLNoTex())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLNoTexUniform())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLSparseBindlessTextureArray())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLSparseBindlessTextureArrayMultiDraw(true))
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLSparseBindlessTextureArrayMultiDraw(false))
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArray())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArrayUniform())
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArrayMultiDraw(true))
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArrayMultiDraw(false))
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArrayMultiDrawBuffer(true))
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsGLTextureArrayMultiDrawBuffer(false))
//            #if WITH_D3D11
//            mSolutions[mProblems.back()->GetName()].push_back(new TexturedQuadsD3D11Naive())
//            #endif
        } else {
            newProb.shutdown()
            newProb.destroy()
        }
    }

    fun getSolutions(problem: Problem, activeApi: OpenGLBase? = null): Array<Solution> {

        val tmpProblems = solutions[problem.name]!!

        activeApi?.let {
            val apiType = activeApi.apiType
            return tmpProblems.filter { it supportsApi apiType }.toTypedArray()
        }

        return tmpProblems
    }
}
